lib/spack/spack/spec.py

```diff
@@ -904,19 +904,29 @@ def dag_hash(self, length=None):
             return b32_hash
 
     def to_node_dict(self):
-        params = dict((name, v.value) for name, v in self.variants.items())
-        params.update(dict((name, value)
-                      for name, value in self.compiler_flags.items()))
+        from .util.spack_yaml import syaml_dict
+
+        ordered_dict = lambda d: syaml_dict(sorted(d.items()))
+
+        params = syaml_dict(sorted(
+            (name, v.value) for name, v in self.variants.items()))
+        params.update(ordered_dict(self.compiler_flags))
         deps = self.dependencies_dict(deptype=('link', 'run'))
-        d = {
+        dependencies = syaml_dict(sorted((
+            (
+                name,
+                ordered_dict({
+                    'hash': dspec.spec.dag_hash(),
+                    'type': [str(s) for s in dspec.deptypes]
```

Comment:
I think we need to sort the deptypes to ensure a stable hash. `'type': sorted([str(s) for s in dspec.deptypes])` should work.